### PR TITLE
Add shopping cart service

### PR DESCRIPTION
**Pull request by blacksam07**
## Summary

- Add Shopping cart service 

## Trello card 

[22-component-communication](https://trello.com/c/AW04E5wB/22-component-communication)

## Screenshots 

shopping-cart-min

### CODE REVIEW

File: src/app/modules/auth/auth.component.scss [hunk 1]
```diff
@@ -28,7 +28,16 @@
   margin-left: 40px;
 }
 
+.actions {
+  width: 150px;
```
Comment (CrisVega08) [inline src/app/modules/auth/auth.component.scss]: sort

File: src/app/modules/auth/auth.component.ts [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 import { Router } from '@angular/router';
+import { ShoppingCartService } from '@app/services/shopping-cart.service';
 
 @Component({
   selector: 'app-auth',
   templateUrl: './auth.component.html',
   styleUrls: ['./auth.component.scss']
 })
 export class AuthComponent implements OnInit {
+  count = 1;
```
Comment (CrisVega08) [inline src/app/modules/auth/auth.component.ts]: why count is setting in 1?

File: src/app/services/shopping-cart.service.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Injectable } from '@angular/core';
+import { BehaviorSubject } from 'rxjs';
+import { Book } from './book';
+import { LocalStorageService } from './local-storage.service';
+
+@Injectable({
+  providedIn: 'root'
```
Comment (CrisVega08) [inline src/app/services/shopping-cart.service.ts]: why in root?

File: src/app/services/shopping-cart.service.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class ShoppingCartService {
+
+  private books = new BehaviorSubject(JSON.parse(this.localStorage.get('shoppingBooks')) || []);
```
Comment (CrisVega08) [inline src/app/services/shopping-cart.service.ts]: I dont like this, i suggest and empty array

**Review by CrisVega08 — CHANGES_REQUESTED** [review]

File: src/app/modules/auth/auth.component.ts (hunk 2 above)
Comment (blacksam07) [reply src/app/modules/auth/auth.component.ts]: Jajajaja sorry 😐 it was a test

File: src/app/modules/auth/auth.component.ts [hunk 5]
```diff
@@ -1,16 +1,19 @@
...
   styleUrls: ['./auth.component.scss']
 })
 export class AuthComponent implements OnInit {
+  count = 0;
 
-  constructor(private router: Router) { }
+  constructor(private router: Router, private shoppingService: ShoppingCartService) { }
 
   ngOnInit(): void {
+    this.shoppingService.shoppingBooks.subscribe(books => this.count = books.length);
```
Comment (CrisVega08) [inline src/app/modules/auth/auth.component.ts]: ok but you need to unsubscribe on component destroy.

File: src/app/modules/auth/auth.module.ts [hunk 6]
```diff
@@ -37,7 +38,8 @@ const routes: Routes = [
       provide: HTTP_INTERCEPTORS,
       useClass: TokenInterceptor,
       multi: true
-    }
+    },
+    ShoppingCartService
```
Comment (CrisVega08) [inline src/app/modules/auth/auth.module.ts:25]: why this?

File: src/app/screens/book-list/book-list.component.html [hunk 7]
```diff
@@ -1,10 +1,11 @@
 <div class="container">
   <input [(ngModel)]="filterValue" #ctrl="ngModel" class="filter" type="text" name="filter" >
   <div class="list-container">
-    <div *ngFor="let book of books | filter:filterValue"  class="list-item" [routerLink]="['/books', book.id]">
-      <a class="item-book-cover"></a>
+    <div *ngFor="let book of books | filter:filterValue"  class="list-item">
```
Comment (CrisVega08) [inline src/app/screens/book-list/book-list.component.html]:
```suggestion
    <div *ngFor="let book of books | filter:filterValue" class="list-item">
```

**Review by CrisVega08 — CHANGES_REQUESTED** [review]

File: src/app/modules/auth/auth.component.html [hunk 8]
```diff
@@ -1,5 +1,11 @@
 <nav class="navbar navbar-nav">
   <a class="logo"></a>
-  <a (click)="logout()"><span class="logout">Logout</span></a>
+  <div class="actions">
+    <a href="" class="shopping-cart">
```
Comment (betoobernal) [inline src/app/modules/auth/auth.component.html]: why do you use the anchor without `href` value? 🤔

File: src/app/screens/book-list/book-list.component.ts [hunk 9]
```diff
@@ -9,12 +10,18 @@ import { BookService } from '@app/services/book.service';
...
   filterValue = '';
+  shoppingBooks;
 
-  constructor(private bookService: BookService) { }
+  constructor(private bookService: BookService, private shoppingService: ShoppingCartService) { }
 
   ngOnInit(): void {
     this.bookService.getBooks()
       .subscribe(({page}) => this.books = page)
+      this.shoppingService.shoppingBooks.subscribe(books => this.shoppingBooks = books);
```
Comment (betoobernal) [inline src/app/screens/book-list/book-list.component.ts]: you should work with all response data at the same time. Check [this](https://rxjs-dev.firebaseapp.com/api/index/function/forkJoin) 👀

File: src/app/services/shopping-cart.service.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+export class ShoppingCartService {
+
+  private books = new BehaviorSubject([]);
+  shoppingBooks = this.books.asObservable();
+
+  constructor() {}
+
+  addBokk (book: Book) {
+    this.books.value.push(book);
+    this.books.next(this.books.value);
```
Comment (betoobernal) [inline src/app/services/shopping-cart.service.ts]:
```suggestion
    const books = [...this.books.value, {...book}];
    this.books.next(books);
```
this is pure and cute.

File: src/app/screens/book-list/book-list.component.ts [hunk 11]
```diff
@@ -9,12 +12,21 @@ import { BookService } from '@app/services/book.service';
...
 
   ngOnInit(): void {
-    this.bookService.getBooks()
-      .subscribe(({page}) => this.books = page)
+    forkJoin(
+      this.bookService.getBooks()
+        .pipe(tap(({page}) => this.books = page)),
+      this.shoppingService.shoppingBooks
+        .pipe(tap(books => this.shoppingBooks = books))
+    ).subscribe()
```
Comment (betoobernal) [inline src/app/screens/book-list/book-list.component.ts]: Nooooope. :sad-harold:

```suggestion
    forkJoin({
      books: this.bookService.getBooks(),
      page: this.shoppingService.shoppingBooks
     }).subscribe(({ books, page })=>{
        this.shoppingBooks = books
        this.books = page
     })
```

File: src/app/screens/book-list/book-list.component.ts [hunk 12]
```diff
@@ -9,12 +12,21 @@ import { BookService } from '@app/services/book.service';
...
   filterValue = '';
+  shoppingBooks;
 
-  constructor(private bookService: BookService) { }
+  constructor(private bookService: BookService, private shoppingService: ShoppingCartService) { }
 
   ngOnInit(): void {
-    this.bookService.getBooks()
-      .subscribe(({page}) => this.books = page)
+    forkJoin(
```
Comment (CrisVega08) [inline src/app/screens/book-list/book-list.component.ts]: you combineLatest operator

**Review by CrisVega08 — APPROVED** [review]

File: src/app/modules/auth/auth.component.ts [hunk 13]
```diff
@@ -1,21 +1,32 @@
...
-export class AuthComponent implements OnInit {
+export class AuthComponent implements OnInit, OnDestroy {
+  count = 0;
+  destroySubject$: Subject<void> = new Subject();
 
-  constructor(private router: Router) { }
+  constructor(private router: Router, private shoppingService: ShoppingCartService) { }
 
   ngOnInit(): void {
+    this.shoppingService.shoppingBooks.pipe(takeUntil(this.destroySubject$)).subscribe(books => this.count = books.length);
```
Comment (betoobernal) [inline src/app/modules/auth/auth.component.ts:19]: indented
```suggestion
    this.shoppingService.shoppingBooks
         .pipe(takeUntil(this.destroySubject$))
         .subscribe(books => this.count = books.length);
```

**Pull request merged**